src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,251 @@ where
+            let bit = path_rev[level];
+            let sibling = auth_path_rev[level];
+
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate_both(
+                two_to_one_hash_parameter,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
+    ///
+    /// Constraints will not be satisfied if root is incorrect.
+    pub fn check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        self.conditionally_check_membership(
+            leaf_hash_parameter,
+            two_to_one_hash_parameter,
+            root,
+            leaf,
+            &Boolean::constant(true),
+        )
+    }
+
+    /// Check membership if `should_enforce` is True.
+    pub fn conditionally_check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+        should_enforce: &Boolean<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        let expected_root =
+            self.calculate_root(leaf_hash_parameter, two_to_one_hash_parameter, leaf)?;
+        // enforce `curr_hash` is root
+        expected_root.conditional_enforce_equal(root, should_enforce);
+        Ok(())
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, constraints will not be satisfied.
+    pub fn update_leaf(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        old_root: &TwoToOneH::OutputVar,
+        old_leaf: &impl ToBytesGadget<ConstraintF>,
+        new_leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        self.conditionally_check_membership(
+            leaf_hash_parameter,
+            two_to_one_hash_parameter,
+            old_root,
+            old_leaf,
+            &Boolean::TRUE,
+        )?;
+        Ok(self.calculate_root(leaf_hash_parameter, two_to_one_hash_parameter, new_leaf)?)
+    }
+
+    /// update the `old_leaf` to `new_leaf` and returned the updated MT root.
+    ///
+    /// If the `old_leaf` does not lead to `old_root`, or `new_leaf` does not lead to `new_root`,
+    /// then constraints will not be satisfied.
+    pub fn check_and_update(
```

Comment:
Should this be called `update_and_check`?

Reply:
yes. updated